src/responder/kcm/kcm_renew.c

```diff
@@ -294,4 +385,165 @@ int kcm_get_renewal_config(struct kcm_ctx *kctx,
     return ret;
 }
 
+static void kcm_renew_tgt_timer_handler(struct tevent_context *ev,
+                                        struct tevent_timer *te,
+                                        struct timeval current_time,
+										void *data)
+{
+    struct renew_tgt_ctx *renew_tgt_ctx = talloc_get_type(data,
+                                                          struct renew_tgt_ctx);
+    errno_t ret;
+    struct timeval next;
+
+    /* forget the timer event, it will be freed by the tevent timer loop */
+    renew_tgt_ctx->te = NULL;
+
+	/* Add any renew-applicable KCM tickets to renew table */
+	ret = kcm_ccdb_renew_init(renew_tgt_ctx->rctx, renew_tgt_ctx->krb5_ctx,
+                              ev, renew_tgt_ctx->db);
+    if (ret != 0) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Failed to add KCM tickets to table.\n");
+        talloc_zfree(renew_tgt_ctx);
+        return;
+    }
+
+    next = tevent_timeval_current_ofs(renew_tgt_ctx->timer_interval, 0);
+    renew_tgt_ctx->te = tevent_add_timer(ev, renew_tgt_ctx,
+                                         next, kcm_renew_tgt_timer_handler,
+                                         renew_tgt_ctx);
+}
+
+
+static void kcm_renew_del_cb(hash_entry_t *entry,
+                             hash_destroy_enum type,
+                             void *pvt)
+{
+    struct renew_data *renew_data;
+
+    if (entry->value.type == HASH_VALUE_PTR) {
+        renew_data = talloc_get_type(entry->value.ptr, struct renew_data);
+        talloc_zfree(renew_data);
+        return;
+    }
 
+    DEBUG(SSSDBG_CRIT_FAILURE,
+          "Unexpected value type [%d].\n", entry->value.type);
+}
+
+errno_t kcm_add_tgt_to_renew_table(struct krb5_ctx *krb5_ctx,
+                                   const char *ccname,
+                                   uid_t uid,
+                                   gid_t gid,
+                                   struct tgt_times *tgtt,
+                                   const char *upn)
+{
+    int ret;
+    hash_key_t key;
+    hash_value_t value;
+    struct renew_data *renew_data = NULL;
+
+    if (krb5_ctx->renew_tgt_ctx == NULL) {
+        DEBUG(SSSDBG_TRACE_LIBS, "Renew context not initialized, "
+                  "automatic renewal not available.\n");
+        return EOK;
+    }
+
+    if (upn == NULL || ccname == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Missing user principal name or ccname.\n");
+        return EINVAL;
+    }
+
+    /* hash_enter copies the content of the hash string, so it is safe to use
+     * discard_const_p here.
+     */
+    key.type = HASH_KEY_STRING;
+    key.str = discard_const_p(char, upn);
```

Comment:
Did you consider using `sss_ptr_hash` interface instead?

Reply:
No I did not, is that interface preferred?

Reply:
It does little bit more work to make sure the hash table stays consistent (does not contain values that were freed, frees values when you delete them, handle duplicate values, etc.) which may or may not be needed in this case. It is also little bit easier to use when you manipulate entries. I would say it is preferred.

At the very least, I don't see you `free(value)` when you `hash_delete(key)`.

Reply:
Thanks for the suggestion, I updated the code to use the `sss_ptr_hash` interface.